File: apps/macos/PasteSheets/Data/DataSources/SettingsDataSource.swift
import Foundation
import SQLite3

protocol SettingsDataSource {
    func get(key: String) throws -> String?
    func set(key: String, value: String) throws
}

final class SettingsDataSourceImpl: SettingsDataSource {
    private let db = DatabaseManager.shared

    func get(key: String) throws -> String? {
        try db.queryOne(
            "SELECT value FROM settings WHERE key = ?1",
            params: [key]
        ) { stmt in
            String(cString: sqlite3_column_text(stmt, 0))
        }
    }

    func set(key: String, value: String) throws {
        try db.execute(
            "INSERT OR REPLACE INTO settings (key, value) VALUES (?1, ?2)",
            params: [key, value]
        )
    }
}

File: apps/windows/PasteSheet/Data/DataSources/SettingsDataSource.cs
using PasteSheet.Data.Database;

namespace PasteSheet.Data.DataSources;

public interface ISettingsDataSource
{
    string? Get(string key);
    void Set(string key, string value);
}

public sealed class SettingsDataSource : ISettingsDataSource
{
    private readonly DatabaseManager _db = DatabaseManager.Shared;

    public string? Get(string key) =>
        _db.QueryOne(
            "SELECT value FROM settings WHERE key = @p1",
            new object?[] { key },
            r => r.GetString(0));

    public void Set(string key, string value) =>
        _db.Execute(
            "INSERT OR REPLACE INTO settings (key, value) VALUES (@p1, @p2)",
            new object?[] { key, value });
}
